documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,720 @@
+/// </summary>
+public interface IDeadzoneHandler
+{
+public float GetValue(IAxisDevice device, int descriptionIndex, Deadzone deadzone);
+public bool IsInDeadzone(IAxisDevice device, int descriptionIndex, Deadzone deadzone);
+
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, Deadzone deadzone);
+    
+    // Requires that the length of the deadzones span matches the number of axes in the group
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, ReadOnlySpan<Deadzone> deadzones);
+}
+```
+
+# Outputs
+Many controllers have haptics, LEDs, etc, that would be nice to access in a generic, normalized way as well. The following is a first draft concept for this functionality that could accomplish this while leaving room for different sorts of outputs later on. 
+
+Feedback requested: is custom data handling necessary here? At that point it violates the "axis" concept pretty strongly and is best served by higher-level interfaces, but I wanted to make sure we at least consider these things.
+
+```csharp
+public partial interface IAxisDevice
+{
+    /// <summary>
+    /// Output definitions for this device - commonly haptics, LEDs, etc
+    /// </summary>
+    IReadOnlyList<OutputDescription> Outputs { get; }
+    
+    /// <summary>
+    /// Allows the caller to set an output value for a specific output index.
+    /// Valid for any axis not marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// </summary>
+    public void SetOutput(int index, float value);
+    
+    /// <summary>
+    /// Allows the caller to set an output value(s) for a non-floating-point output.
+    /// Valid only for axes marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// Requires that the caller knows the correct data type for the output, and as such the implementation must throw
+    /// an exception if the data type is not correct.
+    /// <br/><br/>
+    /// More than likely, concrete implementations or higher-level interfaces will provide a safer way to set these
+    /// sorts of values. The absence of higher-level utilities for this method should be considered a missing feature
+    /// if the implementation utilizes <see cref = "OutputAxisTrait.RawDataOnly"/>.
+    ///<br/><br/>
+    /// It is more or less expected that this will have few, if any, implementations in the early stages of this API.
+    /// Its use is reserved for advanced use-cases or higher-level implementations where the caller knows the exact
+    /// data type required the output - for example, setting a specific bit pattern for a specific LED, driving
+    /// embedded displays, built-in speakers, etc.
+    /// </summary>
+    /// <param name="index">The index of the <see cref="OutputDescription"/> in <see cref="Outputs"/></param>
+    /// <param name="values">A pointer to the value(s) to be set</param>
+    public void SetOutput<T>(int index, T* values, int count = 1) where T : unmanaged => throw new NotImplementedException();
+}

```

Comment:
Come to think of it, we should probably add a `TryGetAxisInfo<T>` and a `TryGetOutputInfo<T>` so that we can have extensions in the future.

Reply:
this could be a nicer way to handle the typing - kind of enforces the idea that axis devices should strive to use similar output types where possible. I wonder how this would extend to the income "OutputGroup"

Reply:
that being said, i think that sort of api needs to be more thoroughly thought out than the next 30 minutes would allow us

Reply:
I think I'd be happy defining these two methods and just having no acceptable types for T atm i.e. Silk.NET reserves the right to add device-type-specific information to aid in axis identification, mapping, configuration, etc. Use of these APIs shouldn't be encouraged as the happy path, but it's sometimes useful to get more info than our intentionally over-generalised API provides.